homeassistant/components/ws66i/media_player.py

```diff
@@ -0,0 +1,213 @@
+        """Handle updated data from the coordinator."""
+        # This will be called for each of the entities after the coordinator
+        # finishes executing _async_update_data()
+
+        # Save a reference to the zone status that this entity represents
+        self._status = self.coordinator.data[self._zone_id_idx]
+        self._set_attrs_from_status()
+
+        # Parent will notify HA of the update
+        super()._handle_coordinator_update()
+
+    @callback
+    def _set_attrs_from_status(self) -> None:
+        status = self._status
+        sources = self._ws66i_data.sources.id_name
+        self._attr_state = STATE_ON if status.power else STATE_OFF
+        self._attr_volume_level = status.volume / float(MAX_VOL)
+        self._attr_is_volume_muted = status.mute
+        self._attr_source = self._attr_media_title = sources[status.source]
+
+    @callback
+    def _async_update_attrs_write_ha_state(self) -> None:
+        self._set_attrs_from_status()
+        self.async_write_ha_state()
+
+    @callback
+    def snapshot(self):
+        """Save zone's current state."""
+        self._snapshot = deepcopy(self._status)
+
+    async def async_restore(self):
+        """Restore saved state."""
+        if not self._snapshot:
+            raise HomeAssistantError("There is no snapshot to restore")
+
+        await self.hass.async_add_executor_job(self._ws66i.restore_zone, self._snapshot)
+        self._status = self._snapshot
+        self._async_update_attrs_write_ha_state()
+
+    async def async_select_source(self, source):
+        """Set input source."""
+        idx = self._ws66i_data.sources.name_id[source]
+        await self.hass.async_add_executor_job(
+            self._ws66i.set_source, self._zone_id, idx
+        )
+        self._status.source = idx
+        self._async_update_attrs_write_ha_state()
+
+    async def async_turn_on(self):
+        """Turn the media player on."""
+        await self.hass.async_add_executor_job(
+            self._ws66i.set_power, self._zone_id, True
+        )
+        self._status.power = True
+        self._async_update_attrs_write_ha_state()
+
+    async def async_turn_off(self):
+        """Turn the media player off."""
+        await self.hass.async_add_executor_job(
+            self._ws66i.set_power, self._zone_id, False
+        )
+        self._status.power = False
+        self._async_update_attrs_write_ha_state()
+
+    async def async_mute_volume(self, mute):
+        """Mute (true) or unmute (false) media player."""
+        await self.hass.async_add_executor_job(
+            self._ws66i.set_mute, self._zone_id, mute
+        )
+        self._status.mute = bool(mute)
+        self._async_update_attrs_write_ha_state()
+
+    async def async_set_volume_level(self, volume):
+        """Set volume level, range 0..1."""
+        await self.hass.async_add_executor_job(
+            self._ws66i.set_volume, self._zone_id, int(volume * MAX_VOL)
```

Comment:
Maybe use `round` instead of `int`.

Reply:
The 2nd parameter of `set_volume()` expects an int. `round()` returns a float. I think keeping it an int is good.